main/src/library/SetMultiMap.flix

```diff
@@ -0,0 +1,514 @@
+}
+
+
+instance Eq[SetMultiMap[k, v]] with Eq[k], Eq[v] {
+    pub def eq(m1: SetMultiMap[k, v], m2: SetMultiMap[k, v]): Bool =
+        SetMultiMap.toList(m1) == SetMultiMap.toList(m2)
+}
+
+instance Order[SetMultiMap[k, v]] with Order[k], Order[v] {
+    pub def compare(x: SetMultiMap[k, v], y: SetMultiMap[k, v]): Comparison =
+        SetMultiMap.toList(x) <=> SetMultiMap.toList(y)
+}
+
+instance Foldable[SetMultiMap[k]] {
+    pub def foldLeft(f: (b, v) -> b & ef, s: b, m: SetMultiMap[k, v]): b & ef = SetMultiMap.foldLeft((acc, v) -> f(acc, v), s, m)
+    pub def foldRight(f: (v, b) -> b & ef, s: b, m: SetMultiMap[k, v]): b & ef = SetMultiMap.foldRight(f, s, m)
+    pub def foldRightWithCont(f: (v, Unit -> b & ef) -> b & ef, s: b, m: SetMultiMap[k, v]): b & ef = SetMultiMap.foldRightWithCont(f, s, m)
+}
+
+instance SemiGroup[SetMultiMap[k,v]] with Order[k], Order[v] {
+    pub def combine(x: SetMultiMap[k,v], y: SetMultiMap[k,v]): SetMultiMap[k,v] = SetMultiMap.union(x, y)
+}
+
+instance CommutativeSemiGroup[SetMultiMap[k,v]] with Order[k], Order[v]
+
+instance Monoid[SetMultiMap[k, v]] with Order[k], Order[v] {
+    pub def empty(): SetMultiMap[k, v] = SetMultiMap.empty()
+}
+
+instance CommutativeMonoid[SetMultiMap[k, v]] with Order[k], Order[v]
+
+namespace SetMultiMap {
+
+    ///
+    /// Returns the empty SetMultiMap.
+    ///
+    pub def empty(): SetMultiMap[k, v] = SetMultiMap(Map.empty())
+
+    ///
+    /// Returns the singleton SetMultiMap where key `k` is mapped to value `v`.
+    ///
+    pub def singleton(k: k, v: v): SetMultiMap[k, v] with Order[k], Order[v] = SetMultiMap(Map.singleton(k, Set.singleton(v)))
+
+    ///
+    /// Returns a SetMultiMap where key `k` is mapped to the singleton set containing `v`.
+    ///
+    pub def fromMap(m: Map[k, v]): SetMultiMap[k, v] with Order[k], Order[v] = SetMultiMap(Map.map(Set.singleton, m))
+
+    ///
+    /// Returns `true` if and only if `m` is the empty SetMultiMap.
+    ///
+    pub def isEmpty(m: SetMultiMap[k, v]): Bool =
+        let SetMultiMap(m1) = m;
+        Map.isEmpty(m1)
+
+    ///
+    /// Returns `Some(vs)` if `k => vs` is in `m`.
+    ///
+    /// Otherwise returns `Nil`.
+    ///
+    pub def get(k: k, m: SetMultiMap[k, v]): Set[v] with Order[k] =
+        let SetMultiMap(t) = m;
+        Map.get(k, t) |> Option.getWithDefault(Set.empty())
+
+    ///
+    /// Returns `true` if and only if `m` contains the key `k`.
+    ///
+    pub def memberOf(k: k, m: SetMultiMap[k, v]): Bool with Order[k] =
+        let SetMultiMap(m1) = m;
+        Map.memberOf(k, m1)
+
+    ///
+    /// Returns the keys of `m`.
+    ///
+    pub def keysOf(m: SetMultiMap[k, v]): Set[k] with Order[k] =
+        let SetMultiMap(m1) = m;
+        Map.keysOf(m1)
+
+    ///
+    /// Returns the values of `m`.
+    ///
+    /// Answer may contain duplicates where values were ascribed to multiple keys.
+    ///
+    pub def valuesOf(m: SetMultiMap[k, v]): List[v] with Order[k] =
```

Comment:
What's the argument that the result should be a list and not say a set, or `List[Set[v]]` or ? :)

Reply:
Not resolved yet.

The argument against `Set` would be different keys to can have common values in the `Set` they point to, so you would lose overall cardinality. But maybe this isn't a bad thing?

Reply:
I see. I don't have an answer. It seems there are multiple plausible paths, but then maybe this one is fine?